lib/src/command/bump.dart

```diff
@@ -0,0 +1,123 @@
+class BumpCommand extends PubCommand {
+  @override
+  String get name => 'bump';
+  @override
+  String get description =>
+      'Updates the version of the current package in pubspec.yaml.';
+  @override
+  String get invocation => 'pub bump [--major|--minor|--patch]';
+
+  bool get isMajor => argResults['major'];
+  bool get isMinor => argResults['minor'];
+  bool get isPatch => argResults['patch'];
+
+  BumpCommand() {
+    argParser.addFlag('major',
+        negatable: false,
+        help: 'Bumps the major version number, while resetting the minor and '
+            'patch numbers, as well as build/pre-release information. '
+            'e.g. 1.2.3 -> 2.0.0');
+    argParser.addFlag('minor',
+        negatable: false,
+        help:
+            'Bumps the minor version number, while resetting the patch number, '
+            'as well as build/pre-release information. e.g. 1.2.3 -> 1.3.0');
+    argParser.addFlag('patch',
+        negatable: false,
+        help: 'Bumps the patch number. e.g. 1.2.3 -> 1.2.4. If it is '
+            'originally a pre-release version, it simply removes the '
+            'pre-release tag instead. e.g. 1.2.3-dev -> 1.2.3');
+  }
+
+  @override
+  void run() {
+    if ((isMajor && isMinor) || (isMajor && isPatch) || (isMinor && isPatch)) {
+      usageException('Only one flag should be specified at most');
+    }
+
+    if (argResults.rest.isNotEmpty && (isMajor || isMinor || isPatch)) {
+      usageException('Must not specify a version to bump to along with flags');
+    }
+
+    final initialVersion = entrypoint.root.version;
+    Version finalVersion;
+
+    if (argResults.rest.isNotEmpty) {
+      if (argResults.rest.length > 1) {
+        usageException('Please specify only one version to bump to');
+      }
+
+      try {
+        finalVersion = Version.parse(argResults.rest.first);
+      } on FormatException {
+        usageException('Invalid version ${argResults.rest.first} found. '
+            'Please ensure that it is correctly formatted according to pub '
+            'semver requirements');
+      }
+    } else {
+      final initialMajor = initialVersion.major;
+      final initialMinor = initialVersion.minor;
+      final initialPatch = initialVersion.patch;
+
+      /// We have already previously ensured that at most one of `isMajor`,
+      /// `isMinor`, or `isPatch` is true.
+      if (isMajor) {
+        finalVersion = Version(initialMajor + 1, 0, 0);
+      } else if (isMinor) {
+        finalVersion = Version(initialMajor, initialMinor + 1, 0);
+      } else {
+        /// Both the `--patch` flag and the default options have the same
+        /// behavior.
+
+        if (initialVersion.isPreRelease) {
+          finalVersion = Version(initialMajor, initialMinor, initialPatch);
+        } else {
+          finalVersion = Version(initialMajor, initialMinor, initialPatch + 1);
+        }
+      }
+    }
+    _updatePubspecVersion(initialVersion, finalVersion);
```

Comment:
I wonder if we should do an acquireDependencies here. Otherwise the next `pub run` command will complain that the pubspec.yaml is newer than the pubspec lock.

Given that we know the dependencies didn't change, we could also consider rewriting the current `pubspec.lock` (and .packages and .dart_tool/package_config.json) to mark them as uop-to-date.

@jonasfj opinions?

Reply:
> I wonder if we should do an acquireDependencies here. Otherwise the next pub run command will complain that the pubspec.yaml is newer than the pubspec lock.

Will it? Won't it just do the verification heuristic to verify that `pubspec.lock` is still a resolution that satisfies `pubspec.yaml`.

Maybe `pub run` will make some noise? I think the assertion checks the timestamps and updates them if there is nothing to be done. And if there is something to be done, then it'll cause a problem. But bumping the version should never causing anything to change (unless you have a cyclic dependency that depends on your specific version of the current package).

Reply:
> Won't it just do the verification heuristic to verify that pubspec.lock is still a resolution that satisfies pubspec.yaml.

Ah yes, that should probably be good enough!